Source/Compositional/FetchedCompositionalDataSource.swift

```diff
@@ -0,0 +1,151 @@
+	public static func `static`(controllers: [SectionType: CompositionalFetchedResultsController], dataSource: UICollectionViewDiffableDataSource<SectionType, ItemType>, delegate: FetchedCompositionalDataSourceDelegate) -> Self {
+		Self.init(controllers: controllers, dataSource: dataSource, delegate: delegate)
+	}
+
+	/// Creates a compositional data source with dynamic sections.
+	/// The sections are provided by the `controller`'s `sectionNameKeyPath` value.
+	public static func dynamic(controller: CompositionalFetchedResultsController, dataSource: UICollectionViewDiffableDataSource<SectionType, ItemType>, delegate: FetchedCompositionalDataSourceDelegate) -> Self {
+		Self.init(controller: controller, dataSource: dataSource, delegate: delegate)
+	}
+
+	private init(controllers: [SectionType: CompositionalFetchedResultsController], dataSource: UICollectionViewDiffableDataSource<SectionType, ItemType>, delegate: FetchedCompositionalDataSourceDelegate) {
+		controllersBySection = controllers.sorted { $0.key.index < $1.key.index }.map { ($0, $1) }
+		self.controllers = controllers.map(\.value)
+		self.dataSource = dataSource
+		self.delegate = delegate
+		super.init()
+		commonInit()
+	}
+
+	private init(controller: CompositionalFetchedResultsController, dataSource: UICollectionViewDiffableDataSource<SectionType, ItemType>, delegate: FetchedCompositionalDataSourceDelegate) {
+		controllersBySection = nil
+		controllers = [controller]
+		self.dataSource = dataSource
+		self.delegate = delegate
+		super.init()
+		commonInit()
+	}
+
+	private func commonInit() {
+		applyEmptySections()
+		setDelegates()
+		performFetches()
+	}
+
+	// MARK: - Controller management
+
+	/// Applies empty sections to secure the order of the sections.
+	private func applyEmptySections() {
+		guard let sections = sections else { return }
+		guard dataSource.snapshot().numberOfSections == 0 else { return }
+		var snapshot: NSDiffableDataSourceSnapshot<SectionType, ItemType> = .init()
+		snapshot.appendSections(sections)
+		dataSource.apply(snapshot, animatingDifferences: false, completion: nil)
+	}
+
+	/// Sets the delegates of all NSFetchedResultsControlle instances.
+	private func setDelegates() {
+		controllers.forEach { $0.set(delegate: self) }
+	}
+
+	/// Executes the fetch requests of all NSFetchedResultsController instances.
+	public func performFetches() {
+		do {
+			try controllers.forEach { try $0.performFetch() }
+		} catch let error {
+			assertionFailure("Error performing controller fetch: \(error)")
+		}
+	}
+
+	// MARK: - NSFetchedResultsControllerDelegate
+
+	public func controllerWillChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
+		delegate.willChangeContent()
+	}
+
+	public func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChangeContentWith snapshot: NSDiffableDataSourceSnapshotReference) {
```

Comment:
From my testing (and some posts found on the Apple dev forum) it turns out that object updates are covered when building with iOS SDK 15. So no extra code was needed to cover object updates.